File: implementations/dgs/src/main/java/com/netflix/graphql/dgs/compatibility/model/DeprecatedProduct.java
package com.netflix.graphql.dgs.compatibility.model;

public class DeprecatedProduct {
    public static DeprecatedProduct DEPRECATED_PRODUCT = new DeprecatedProduct("apollo-federation-v1", "@apollo/federation-v1", "Migrate to Federation V2");

    private final String sku;
    private final String pkg;
    private final String reason;
    private final User createdBy;
    public DeprecatedProduct(String sku, String pkg, String reason) {
        this.sku = sku;
        this.pkg = pkg;
        this.reason = reason;
        this.createdBy = User.DEFAULT_USER;
    }

    public String getSku() {
        return sku;
    }

    public String getPackage() {
        return pkg;
    }

    public String getReason() {
        return reason;
    }

    public User getCreatedBy() {
        return createdBy;
    }

    public static DeprecatedProduct resolveBySkuAndPackage(String sku, String pkg) {
        if (DEPRECATED_PRODUCT.sku.equals(sku) && DEPRECATED_PRODUCT.pkg.equals(pkg))  {
            return DEPRECATED_PRODUCT;
        } else {
            return null;
        }
    }
}

File: implementations/hotchocolate/Types/DeprecatedProduct.cs
using ApolloGraphQL.HotChocolate.Federation;

namespace Products;

[Key("sku package")]
public class DeprecatedProduct
{
    public DeprecatedProduct(string sku, string package, string? reason, User? createdBy)
    {
        Sku = sku;
        Package = package;
        Reason = reason;
        CreatedBy = createdBy;
    }

    public string Sku { get; }

    public string Package { get; }

    public string? Reason { get; }

    public User? CreatedBy { get; }

    [ReferenceResolver]
    public static DeprecatedProduct? GetProductByPackage(
        string sku,
        string package,
        Data repository)
        => repository.DeprecatedProducts.FirstOrDefault(
            t => t.Sku.Equals(sku) &&
                t.Package.Equals(package));
}
